src/sources/geniuscom.cpp

```diff
@@ -71,57 +75,112 @@ std::vector<LyricDataRaw> GeniusComSource::search(const LyricSearchParams& param
+            result.artist = search_artist->valuestring;
+
+			song_metadata.push_back(result);
         }
+
+        cJSON_Delete(json);
     }
 
-    if(lyric_text.empty())
+    return song_metadata;
+}
+
+bool GeniusComSource::lookup(LyricDataRaw& data, abort_callback& abort)
+{
+    auto request = http_client::get()->create_request("GET");
+    request->add_header(API_KEY_HEADER);
+
+    std::string url = "https://api.genius.com" + std::string(data.lookup_id) + "?text_format=plain";
+
+    LOG_INFO("Page %s retrieved", url.c_str());
+    std::string lyric_text;
+
+    pfc::string8 content;
+    try
     {
-        throw new std::runtime_error("Failed to parse lyrics, the page format may have changed");
+        file_ptr response_file = request->run(url.c_str(), abort);
+        response_file->read_string_raw(content, abort);
+        // NOTE: We're assuming here that the response is encoded in UTF-8 
     }
-    else
+    catch (const std::exception& e)
     {
-        LOG_INFO("Successfully retrieved lyrics from %s", url.c_str());
-        const std::string_view trimmed_text = trim_surrounding_whitespace(lyric_text);
-
-        LyricDataRaw result = {};
-        result.source_id = id();
-        result.source_path = url;
-        result.artist = params.artist;
-        result.album = params.album;
-        result.title = params.title;
-        result.type = LyricType::Unsynced;
-        result.text_bytes = string_to_raw_bytes(trimmed_text);
-        return {std::move(result)};
+        LOG_WARN("Failed to download genius.com page %s: %s", url.c_str(), e.what());
+        return {};
     }
-}
 
-bool GeniusComSource::lookup(LyricDataRaw& /*data*/, abort_callback& /*abort*/)
-{
-    LOG_ERROR("We should never need to do a lookup of the %s source", friendly_name().data());
-    assert(false);
+    LOG_INFO("Successfully retrieved lyrics from %s", url.c_str());
+
+    {   // Parser gets its own scope
+        cJSON* json = cJSON_Parse(content.c_str());
+        if (json == nullptr) {
+            LOG_WARN("Failed to download genius.com page %s: %s", url.c_str(), "Failed to parse JSON");
+            cJSON_Delete(json);
+            return {};
+        }
+
+        const cJSON* song_response = cJSON_GetObjectItem(json, "response");
+        const cJSON* song_song = cJSON_GetObjectItem(song_response, "song");
+        const cJSON* song_lyrics = cJSON_GetObjectItem(song_song, "lyrics");
+        const cJSON* song_lyrics_plain = cJSON_GetObjectItem(song_lyrics, "plain");
+        const cJSON* song_title = cJSON_GetObjectItem(song_song, "title");
+        const cJSON* song_album = cJSON_GetObjectItem(song_song, "album");
+
+        if (song_lyrics_plain == nullptr) { // fart
+            LOG_WARN("Failed to download from genius.com page: No lyrics data!");
+            cJSON_Delete(json);
+            return {};
+        }
+
+        if (song_title->valuestring) { data.title = song_title->valuestring; };

```

Comment:
We already have an artist/album/title from the search and then we've looked up exactly that lyric by ID here, so I don't think we need to pull out the other metadata again, we just need to populate the text.